src/QMCDrivers/DMC/WalkerControlMPI.cpp

```diff
@@ -390,4 +447,163 @@ void WalkerControlMPI::swapWalkersSimple(MCWalkerConfiguration& W)
+                << "neither good nor bad, exist on a node. "
+                << "Please report to developers. " << std::endl;
+    APP_ABORT("WalkerControlMPI::swapWalkersSimple no existing walker");
+  }
+
+  if (plus.size() != minus.size())
+  {
+    app_error() << "Walker send/recv pattern doesn't match. "
+                << "The send size " << plus.size() << " is not equal to the receive size " << minus.size() << " ."
+                << std::endl;
+    throw std::runtime_error("Trying to swap in WalkerControlMPI::swapWalkersSimple with mismatched queues");
+  }
+
+  UPtrVector<MCPWalker> new_walkers;
+  std::vector<int> copies_of_each_new_walker;
+
+  // sort good walkers by the number of copies
+  std::vector<std::pair<int, std::reference_wrapper<MCPWalker>>> sorted_good_walkers;
+  for (int iw = 0; iw < ncopy_w.size(); iw++)
+    sorted_good_walkers.push_back(std::make_pair(adjust.copies_to_make[iw], adjust.good_walkers[iw]));
+  // Sort only on the number of copies
+  std::sort(sorted_good_walkers.begin(), sorted_good_walkers.end(),
+            [](auto& a, auto& b) { return a.first > b.first; });
+
+  
+  int nsend = 0;
+
+  int nswap = plus.size();
+  std::vector<WalkerMessage> message_list(nswap);
+  int messages_incoming = 0;
+  for (int ic = 0; ic < nswap; ic++)
+  {
+    if (plus[ic] == MyContext)
+    {
+      // always send the last good walker
+      message_list.push_back(WalkerMessage{sorted_good_walkers.back().second, minus[ic]});
+      --(sorted_good_walkers.back().first);
+      if (sorted_good_walkers.back().first == 0)
+        sorted_good_walkers.pop_back();
+    }
+    else if(minus[ic] == MyContext)
+    {
+      ++messages_incoming;
+      if ( adjust.bad_walkers.size() > 0)
+      {
+        pop.killWalker(adjust.bad_walkers.back());
+        adjust.bad_walkers.pop_back();
+      }
+    }
+            
+
+  }
+
+  std::vector<mpi3::request> send_requests;
+  if( message_list.size() > 0)
+  {
+    std::queue<WalkerMessage> message_queue;
+    auto it_walker_messages = message_list.begin();
+    while (it_walker_messages != message_list.end())
+    {
+      if ((it_walker_messages+1 != message_list.end()) && (*(it_walker_messages + 1) == *it_walker_messages))
+        (it_walker_messages+1)->multiplicity++;
+      else
+        message_queue.push(*it_walker_messages);
+   
+      ++it_walker_messages;
+    }
+    
+    while ( !message_queue.empty() )
+    {
+      WalkerMessage message = message_queue.front();
+      message_queue.pop();
+      send_requests.push_back(myComm->comm.isend_value(message.multiplicity, message.target_rank));
+      size_t byteSize    = message.walker.byteSize();
+      send_requests.push_back(myComm->comm.isend_n(message.walker.DataSet.data(), byteSize, message.target_rank));
+    }
+  }                                                    
+
+  std::vector<mpi3::request> receive_requests;
+  messages_incoming *= 2; // extra message sent for multiplicity
```

Comment:
What is the magic number 2 here?

Reply:
That was for the separate number of copies message, the next push to this PR will remove this.